Name-suggestion-index v6

bhousel opened the pull request:
This updates the iD code to support the latest name-suggestion-index.  I'll release v5 in the next few days, but I have tested this locally and works awesome.

Some enhancements:
- The presets and fields have been upgraded to fully support locationSets now
- Added a `coreLocations` module to resolve and cache locationSets used anywhere in iD (presets, community-index, name-suggestions).  Processing happens in chunks during browser idle callbacks.
- NSI presets are downloaded and merged in after iD starts up. The presets become available within a few seconds. 
- Upgraded parts of the code (field locking, map pin styling) that previously only supported `brand:wikidata` to now support the other kinds of features in NSI.  (flags, operators, transit networks)
- Upgraded the outdated tags and suspicious tags validators to use the latest NSI data and matcher.  
- Note about the validator: there is an interesting issue here in that the first few tiles downloaded from OSM won't be fully validated because the NSI data isn't ready yet.  This was always an issue before but it is a bit more annoying now.  I'd like to Promisify all the validators so they just run whenever they are ready to run (update: solved it in #8319 and merged it here)

I'll have to look and see what outstanding issues this closes, but it's all the ones I know about.

re: https://github.com/osmlab/name-suggestion-index/issues/4543
closes #8426
closes #8304
closes #8271 
closes #8237 
closes #8143
closes #6890
closes #6055
closes #6408
closes #5484

bhousel committed:
Update to name-suggestion-index v5
Merge in NSI presets after UI has initialized

bhousel committed:
Update presetIndex to resolve and index locationSets
Added a deferred work queue to do this with requestIdleCallback in the background

bhousel committed:
Move the location index and resolver into a global coreLocations

bhousel committed:
Pre-resolve the world locationSet

bhousel committed:
Use locationManager to filter fields/presets/defaults

bhousel committed:
Use locationManager to resolve/query community index

bhousel committed:
Have mergeLocationSets work on Objects, add locationSetID property
This means less work for the other parts of the code,
now they don't need to chain that extra step after the promise.

bhousel committed:
coreLocation tests, documentation

bhousel committed:
Upgrade validations outdated_tags, suspicious_names to NSI v5

bhousel committed:
NSI match returns an object now

bhousel committed:
Actually assign the locationSetID properties on the given objects
(should have been in 868db8819)

bhousel committed:
Make sure suggestion preset terms get used in the search

bhousel committed:
Support more `*:wikidata` tags for field locking and pin styling

1ec5 commented on modules/validations/outdated_tags.js:
```diff
@@ -17,58 +16,126 @@ let _nsi;
+  // Returns true if this tag key is a "namelike" tag that the NSI matcher would have indexed..
+  function isNamelike(k) {
+    const namePatterns = [
+      /^(flag:)?name$/i,                                             // e.g. `name`, `flag:name`
```
openstreetmap/id-tagging-schema#114 requests a `flag:name` field for the Flagpole preset.

1ec5 commented on modules/validations/outdated_tags.js:
```diff
@@ -17,58 +16,126 @@ let _nsi;
+      // There are a few exceptions to the namelike regexes.
+      // Usually a tag suffix contains a language code like `name:en`, `name:ru`
+      // but we want to exclude things like `operator:type`, `name:etymology`, etc..
+      if (/:(colour|type|left|right|etymology|wikipedia)$/.test(k)) return false;
```
`name:pronunciation` is another common subkey of `name` that isn’t a language code. That said, IPA is very unlikely to match a name in NSI anyways.

bhousel replied:
Yeah I _think_ that's ok - This code is just collecting values to attempt to match, so if collects a few extras it's probably fine, but will waste a few microseconds.

bhousel replied:
I removed it anyway in a32553595

1ec5 commented on modules/validations/outdated_tags.js:
```diff
@@ -79,7 +146,7 @@ export function validationOutdatedTags() {
+    // Add missing addTags from the detected preset
     let newTags = Object.assign({}, entity.tags);  // shallow copy
     if (preset.tags !== preset.addTags) {
       Object.keys(preset.addTags).forEach(k => {
```
Each flag entry in NSI comes with a `flag:name` tag, but that’s mainly for internal reference. The `flag:name` key would logically be in the local language, whereas the value in NSI is specific to English, so it wouldn’t be appropriate to add the `flag:name` from NSI in non-English-speaking countries. (Apologies if the code below already handles that; I couldn’t tell.)

bhousel replied:
> (Apologies if the code below already handles that; I couldn’t tell.)

It doesn't but it should - this is a good catch..

1ec5 commented on modules/validations/outdated_tags.js:
```diff
@@ -93,63 +160,120 @@ export function validationOutdatedTags() {
+          // Exceptions, throw out the match
+          if (
+            (!itemQID || itemQID === notQID) ||       // no `*:wikidata` or matched a `not:*:wikidata`
+            (newTags.office && !item.tags.office)     // feature may be a coprorate office for a brand? - #6416
```
Would a corporate office still be a decent candidate for an `operator` preset?

bhousel replied:
I'm not sure.. I kind of think no, except in cases where it's really widespread or like a chain.  (Military recruiting centers maybe or DMVs?)

I suppose any business can also have a corporate office.  Do people tag them like `office=yes`?  Adding an extra preset for each one seems like it would just confuse users.

For reference, these are the brand-like `office=*` categories NSI tracks today:
<img width="300" alt="Screen Shot 2021-01-18 at 7 17 31 PM" src="https://user-images.githubusercontent.com/38784/104973139-da3e2c80-59c1-11eb-9ebd-9130663c3152.png">

1ec5 commented on modules/ui/fields/input.js:
```diff
@@ -34,11 +34,41 @@ export function uiFieldText(field, context) {
+                var isSuggestion = preset && preset.suggestion;
+
+                // Lock the field if there is a value and a companion `*:wikidata` value
+                return isSuggestion && !!entity.tags[which] && !!entity.tags[which + ':wikidata'];
```
`flag:name` goes with `flag:wikidata` rather than `flag:name:wikidata`. (Sorry, I didn’t think of the inconsistency when promoting `flag:name` instead of `flag`.)

Maybe this is a happy accident, since the mapper may need to come up with the flag name in the local language anyways. Also, for some kinds of flags, it’s common enough for the same flag to represent a different entity and be known by a different name. (The example that comes to mind is the flag of the Vatican City, which is also a religious flag representing the Holy See, but there are also plenty of examples among historical flags.)

bhousel replied:
> Maybe this is a happy accident, since the mapper may need to come up with the flag name in the local language anyways.

Yes it is 👍   It means this code would lock a `flag` field, but not a `flag:name` field, which is what I think you want..

1ec5 reviewed (commented):
Exciting to see more parts of NSI become accessible to ordinary mappers. This is the first time I’m looking at the matcher code, so it’s possible that you’ve already accounted for some of the feedback I’ve given below.

bhousel commented on modules/core/file_fetcher.js:
```diff
@@ -19,8 +19,14 @@ export function coreFileFetcher() {
-    'nsi_brands': 'https://cdn.jsdelivr.net/npm/name-suggestion-index@4/dist/brands.min.json',
-    'nsi_filters': 'https://cdn.jsdelivr.net/npm/name-suggestion-index@4/dist/filters.min.json',
+
+    'nsi_presets': 'https://raw.githubusercontent.com/osmlab/name-suggestion-index/main/dist/presets/nsi-id-presets.min.json',
```
Oh I should add - I don't actually want to load these from the GitHub `main` branch.. 
These URLs just for testing it.

bhousel commented on modules/core/locations.js:
```diff
@@ -0,0 +1,267 @@
+  //   …
+  //  ]
+  //
+  //  Returns a Promise fullfilled when the resolving/indexing has been completed
```
haha spellchecker

bhousel commented on modules/validations/outdated_tags.js:
```diff
@@ -17,58 +16,126 @@ let _nsi;
+    });
+  }
+
+  // This Promise will fullfill after NSI presets are loaded and locations merged into the locationManager.
```
Well at least my misspelling is consistent.

bhousel commented on modules/validations/outdated_tags.js:
```diff
@@ -93,63 +160,120 @@ export function validationOutdatedTags() {
+          subtype = 'noncanonical_brand';
+
+          // Preserve some tags values that we don't want NSI to overwrite.
+          const keepTags = ['takeaway', 'building']
```
@1ec5 we should add `flag:name` here to leave people's local flag names alone.

bhousel replied:
I did this in f61a3ef5e, and it works:

before | after
--- | ---
<img width="350" alt="Screen Shot 2021-01-18 at 9 37 04 PM" src="https://user-images.githubusercontent.com/38784/104981433-9a357480-59d6-11eb-998a-5c4b8126f355.png"> | <img width="350" alt="Screen Shot 2021-01-18 at 9 42 27 PM" src="https://user-images.githubusercontent.com/38784/104981447-a0c3ec00-59d6-11eb-859a-d76eb0403003.png">

bhousel committed:
Fix teh misspellings

tas50 commented:
@bhousel Any chance you can push v5 so this can be tested locally?

bhousel commented:
> @bhousel Any chance you can push v5 so this can be tested locally?

What I've been doing for now is using [`npm link`](https://dev.to/erinbush/npm-linking-and-unlinking-2h1g) to test locally:
```bash
cd path/to/name-suggestion-index
npm link
cd path/to/iD
npm link name-suggestion-index
```

This makes a symlink between the two projects, which is great for local development.. 
But as soon as I'm really sure we won't need to change anything else about NSI's matcher I'll release it.

bhousel committed:
`name:pronunciation` is not namelike

bhousel committed:
Don't replace `flag:name` tag - it's expected to be in local language

tas50 commented:
@bhousel That is some JS magic right there. You're making me sad to be in Ruby land. I was able to fire up the branch just fine with link. It does look like some places are too aggressive suggesting name changes. That feels like the iD side of things.

Here's a Portland independent car repair shop that's tagged as repairing Volvos with brand:volvo

<img width="391" alt="Screen Shot 2021-01-18 at 7 52 00 PM" src="https://user-images.githubusercontent.com/1015200/104986885-0f9a4880-59c9-11eb-856a-1f45f15c3690.png">

Here's a hotel with a custom name set:

<img width="384" alt="Screen Shot 2021-01-18 at 7 51 35 PM" src="https://user-images.githubusercontent.com/1015200/104986900-1fb22800-59c9-11eb-914c-cc6b83ead647.png">

This one is real strange, but it made me chuckle:

<img width="390" alt="Screen Shot 2021-01-18 at 7 52 31 PM" src="https://user-images.githubusercontent.com/1015200/104986919-335d8e80-59c9-11eb-9dc9-2351ff650586.png">

bhousel commented:
> @bhousel That is some JS magic right there. You're making me sad to be in Ruby land. I was able to fire up the branch just fine with link. It does look like some places are too aggressive suggesting name changes. That feels like the iD side of things.

Thanks @tas50 - these are great examples.. Let me know if you see any more like this.
There is some logic in the field UI for "whether the name field should be readonly/locked" and I think I can put some similar logic into the validator for "whether the name tag should be replaced with what is in NSI".

I'm really curious about what's going on with the "Vacant" Applebees. Is it using a tag like `name:old` or `name:disused`? 🤔  
If you have the node/way id, I'd like to take a look!

tas50 commented:
@bhousel It turns out the Applebees was decom'd wrong and had the official_name tag left so that sort of makes sense. I found a few more places where things are a bit too aggressive overwriting name fields:

Wireless stores that have a brand set. I think these are sort of mistagged, but that's a whole different issue. They're basically Verizon stores with an operator of XYZ, but they get tagged as XYZ with the brand of Verizon.
<img width="379" alt="Screen Shot 2021-01-18 at 10 12 10 PM" src="https://user-images.githubusercontent.com/1015200/104996915-92c59980-59dd-11eb-8a3f-9f4859932294.png">

bhousel committed:
Preserve `name` value if this preset shows `brand` or `operator` field

This also fixes the logic for calculating whether the preset shows a
`brand` or `operator` field - it needs to use `fields()` to actually
resolve the fields, as these fields can be inherited from another preset.

This also includes a change to match "primary" names before
"alternate" names (aka the "Baby Gap" / "Gap" problem)

1ec5 commented:
> Wireless stores that have a brand set. I think these are sort of mistagged, but that's a whole different issue. They're basically Verizon stores with an operator of XYZ, but they get tagged as XYZ with the brand of Verizon.

It’s probably one of those Verizon Authorized Retailers where either the store name or Verizon could have top billing. It’s hard to say what other POI categories tend to have store-specific names; it probably varies from country to country.

bhousel commented:
@tas50 & @1ec5  I just pushed a change that should make the name overwriting much less aggressive.

Basically the rule works the same as the way it has worked before for Name field locking 🔒 :
* if the preset _does_ display a "Brand" field, the Brand field gets locked and the Name field is editable.  NSI _will not_ overwrite the name tag.
* if the preset _does not_ display a "Brand" field, the Name field is locked.  NSI _will_ overwrite the name tag.
* (same rules above apply for presets that use "Operator" instead of "Brand")

This means that we can control which categories we expect mappers to supply store-specific names (Car Dealerships, Hotels, Post Offices, etc).  AFAIK the presets already set up correctly, but we can adjust in the [id-tagging-schema](https://github.com/openstreetmap/id-tagging-schema) repo if needed..

UKChris-osm commented:
Some great with this @bhousel thank you 👍 

How's it looking with adding an operator within a location, such as an Ambulace Station in the [South West of England](https://ideditor.github.io/location-conflation/?locationSet=%7B%22include%22%3A%5B%22gb-south-west.geojson%22%5D%7D&referrer=nsi)?

Does placing a node / way in the area suggest [South Western Ambulance Service](https://nsi.guide/index.html?t=operators&k=emergency&v=ambulance_station#southwesternambulanceservice-bba5ce) as expected?

bhousel commented:
> How's it looking with adding an operator within a location, such as an Ambulace Station in the [South West of England](https://ideditor.github.io/location-conflation/?locationSet=%7B%22include%22%3A%5B%22gb-south-west.geojson%22%5D%7D&referrer=nsi)?
> Does placing a node / way in the area suggest [South Western Ambulance Service](https://nsi.guide/index.html?t=operators&k=emergency&v=ambulance_station#southwesternambulanceservice-bba5ce) as expected?

Yes, it is working wonderfully!  
Thanks for your help in contributing POIs for this region to the index 🙇  

I see this when I type "south" around **Plymouth, South West England**:

<img width="500" alt="Screen Shot 2021-01-20 at 3 24 11 PM" src="https://user-images.githubusercontent.com/38784/105230724-ca465a00-5b33-11eb-96a4-cb7d57a49f30.png">

I see this when I try it around **Cardiff, Wales**:

<img width="600" alt="Screen Shot 2021-01-20 at 3 27 38 PM" src="https://user-images.githubusercontent.com/38784/105230949-1beee480-5b34-11eb-93e9-31220385a62d.png">

UKChris-osm commented:
Fantastic :)

I've enjoyed contributing :)

What happens if you place a node in the South West but start typing "_Ambulance_" for example, will it suggest a branded operator Ambulance Station as the first result, or does the generic Ambulance Station POI within iD still dominate?

bhousel commented:
> What happens if you place a node in the South West but start typing "_Ambulance_" for example, will it suggest a branded operator Ambulance Station as the first result, or does the generic Ambulance Station POI within iD still dominate?

<img width="976" alt="Screen Shot 2021-01-20 at 4 18 59 PM" src="https://user-images.githubusercontent.com/38784/105241147-43957b00-5b3b-11eb-9ece-bb4ce976d5de.png">

It matches the generic presets first, but the branded ones rank pretty highly..
Here's what I see in Cornwall - does this look reasonable?
(I'm sure we can scope the locationSets a bit better, but at least we're not seeing anything from other countries in this list)

UKChris-osm commented:
Yes, looks like it's working great 👍 

I am hoping to add a few more .geojson's tomorrow / Friday as the first two items (North and East) aren't based in Cornwall, but I expect them to appear at the moment as they are "gb-eng" so shouldn't appear in, say, Wales.

Also, I guess pressing the "i" button next to each one above would display a bit of text from Wikipedia / Wikidata?

bhousel committed:
Match the prereleased nsi v5 (for now)

UKChris-osm commented:
> Here's what I see in Cornwall - does this look reasonable?

I've covered England with some .geojson files for each region, and assigned police / fire / ambulance stations to these, so the locationing should be better now :)

peternewman commented on modules/validations/outdated_tags.js:
```diff
@@ -93,63 +158,148 @@ export function validationOutdatedTags() {
+          // Exceptions, throw out the match
+          if (
+            (!itemQID || itemQID === notQID) ||       // no `*:wikidata` or matched a `not:*:wikidata`
+            (newTags.office && !item.tags.office)     // feature may be a coprorate office for a brand? - #6416
```
Typo:
```suggestion
            (newTags.office && !item.tags.office)     // feature may be a corporate office for a brand? - #6416
```

PRed into Codespell so would flag up eventually...

bhousel replied:
fixed it in 3640e15fa, thanks for catching 👍

peternewman replied:
Thanks!

bhousel committed:
Rewrite the validator in ES6/Promises, several improvements here:
- implements a validation work queue, jobs are run during browser idle callbacks
- when merging base entities, don't run validations 2x on both base and head graphs (this was wasteful)
- keep track of resolved issues in a separate set (it's not a simple compare of base/head anymore)
  this happens after validation queue is empty and avoids race conditions and inaccurate resolved counts

tas50 commented:
@bhousel It looks like the new flag pole logic is a bit aggressive as well:

Flag pole with multiple flags defined:
![image](https://user-images.githubusercontent.com/1015200/105799505-8dba9a00-5f49-11eb-9ecf-a2dad2a2fe6f.png)

Coast Guard signal flag:
<img width="398" alt="image" src="https://user-images.githubusercontent.com/1015200/105799544-a7f47800-5f49-11eb-9ce6-7f76975612ee.png">

bhousel committed:
Allow validators to return provisional results, revalidate after delay
Also add a ton of commments to validator.js

bhousel committed:
Merge pull request #8319 from openstreetmap/promisify_validation

Promisify validation

bhousel committed:
Fix misspelling "coprorate" -> "corporate"

bhousel commented:
> It looks like the new flag pole logic is a bit aggressive as well:

Thanks for catching these @tas50 .. I'll adjust how the code handles flagpoles (this is all new and we're still learning).

> Flag pole with multiple flags defined:

I think this one happened because there is a `country=US` tag.  
The [osm wiki](https://wiki.openstreetmap.org/wiki/Tag:man_made%3Dflagpole) says: "For national flags to indicate which country is represented on the flag (examples: United States of America, Germany, Tanzania)." so it's a bit incorrect to add this tag to a flagpole with multiple flags on it, but I can see why people do it and the wiki is just a suggestion anyway.  I'll add some code to 1) skip any feature where the namelike tag contains semicolons, 2) consider `country` after other values.  

My concern here is that people may tag things like `flag:name=NJ`+`country=US` which is again - kind of wrong but I can see why people would do it - and we don't want to suggest an "upgrade" to the United States national flag.  So I have to think about this more.

> Coast Guard signal flag:

This one probably happened because there is an `operator=` tag with a value like "US Coast Guard" or "USCG" or something.  On the NSI side we have a hierarchy where things are placed into "trees" and there is better logic for knowing which tags matter for which trees.  So I need to push this ability into the iD code too.

bhousel committed:
Adust NSI matching validation code:
- don't try matching semicolon-separated lists to NSI
- better matching for flagpoles with a `country` tag
- better lists of namelike keys (don't match flag operator)

see: https://github.com/openstreetmap/iD/pull/8305#issuecomment-769174070

bhousel committed:
Also include `operator:wikidata` as a wikidata tag
osmEntity#hasWikipedia was only used by the suspicious_name validator

1ec5 commented:
>> I think this one happened because there is a `country=US` tag.  
> 
> The [osm wiki](https://wiki.openstreetmap.org/wiki/Tag:man_made%3Dflagpole) says: "For national flags to indicate which country is represented on the flag (examples: United States of America, Germany, Tanzania)." so it's a bit incorrect to add this tag to a flagpole with multiple flags on it, but I can see why people do it and the wiki is just a suggestion anyway.  I'll add some code to 1) skip any feature where the namelike tag contains semicolons, 2) consider `country` after other values.  

Yeah, there’s no good answer to this one. `country=US` is correct for one part of `flag:wikidata` but there’s no correct value for the other part. In theory, we could come up with a lanes-like syntax so that `country=US|` would correspond to `flag:wikidata=Q42537|Q330517`, but the whole point of continuing to use `country` on flagpoles is graceful degradation in existing renderers like OSM2World (which doesn’t understand multiple flags). If the validator skips features with `flag:wikidata` set to multiple values, then I think that’s good enough to sidestep the larger issue of arrays-of-key-value-stores for now.

bhousel committed:
Remove the brand combo from the name field
This has caused problems going back a long time, and now that we have a validator
and very visible brand presets with logos, it's not necessary anymore.
(re: #5558, closes #8271, closes #8304)

bhousel committed:
Better handling of headGraph, separate head and base queues
This involves a few things to make the validator less weird
- _headGraph shouldn't be allowed to change while validation is happening..
- So we don't allow that to happen anymore, and keep track of _headPromise and _headIsCurrent
- If head graph falls behind, kick off another validation to catch it up
- Separate head and base work queues, so we aren't waiting for the base entities to validate
  before providing feedback to the user about what they are editing
  (the base queue can get quite large around metropolitan areas)

bhousel committed:
Checkin en.min.json

tas50 commented:
I found another weird operator suggestion @bhousel where an emergency room tag is being removed

<img width="375" alt="Screen Shot 2021-02-13 at 5 54 28 PM" src="https://user-images.githubusercontent.com/1015200/107866375-129c2380-6e25-11eb-9b09-e28a9c24cbb2.png">

tas50 commented:
Amazon Lockers are also a bit tricky since each one has a custom name and it's handy to know that when you schedule delivery

<img width="381" alt="image" src="https://user-images.githubusercontent.com/1015200/107866794-45e0b180-6e29-11eb-99ef-35a175af0209.png">

peternewman commented:
> Amazon Lockers are also a bit tricky since each one has a custom name and it's handy to know that when you schedule delivery
> 
> <img alt="image" width="381" src="https://user-images.githubusercontent.com/1015200/107866794-45e0b180-6e29-11eb-99ef-35a175af0209.png">

Normally, like Weatherspoons pubs, we'd just remove the name in the preset:
https://github.com/osmlab/name-suggestion-index/blob/504a8a2a306abec8cba118a9f90c35a297cae765/data/brands/amenity/vending_machine.json#L49

Which will stop it overwriting.

At first I thought @bhousel was suggesting something smarter would happen here:
> * if the preset _does_ display a "Brand" field, the Brand field gets locked and the Name field is editable.  NSI _will not_ overwrite the name tag.

But I think that relies on the brand field being shown in the Parcel Pickup Locker preset (which perhaps it should).

I wonder if the other name "Hemlock" should be the ref perhaps?

Adamant36 commented:
> I wonder if the other name "Hemlock" should be the ref perhaps?

That would be my guess. Or something else since the ref key is really more to do with numbers (I'd be extremely surprised if there were not legitimate reference numbers for Amazon Lockers that should be in the ref field and not just location names).

1ec5 commented:
[`branch`](https://wiki.openstreetmap.org/wiki/Key:branch) might be more appropriate than `ref`. I don’t have a strong opinion on where to put that information for Amazon lockers, but sometimes a `name` or `official_name` is the only way a data consumer knows to concatenate `brand`, `ref`, `branch`, etc. into a single coherent, human-readable name. It’s common to give post offices individual branch names, so it makes sense to me that a non-office delivery point might be named similarly.

bhousel commented:
>Amazon Lockers are also a bit tricky since each one has a custom name and it's handy to know that when you schedule delivery

Good catch @tas50 - I put some thoughts here on the Green King PR, as this is a thing that keeps coming up and I have to figure out:
https://github.com/osmlab/name-suggestion-index/issues/4906
tl;dr - NSI should keep track of whether to apply the name tag as a default only, or stictly (overwrite what's there).

> I found another weird operator suggestion @bhousel where an emergency room tag is being removed

Ugh, also good catch, and I know why this is.. It's because `emergency=*` is also used as a toplevel/defining tag for things like ambulance and rescue stations, see here: https://github.com/osmlab/name-suggestion-index/tree/main/data/operators/emergency
When OSM uses a single tag in multiple ways, stuff always gets weird.

I'll fix these..

peternewman commented:
> That would be my guess. Or something else since the ref key is really more to do with numbers (I'd be extremely surprised if there were not legitimate reference numbers for Amazon Lockers that should be in the ref field

I'm sure somewhere in their database their is a more standardised identifier, but they seem to work really hard to hide it on their website. Probably because people can't accidentally mix up a name as 1243 for example. Similar to What 3 Words.

> and not just location names).

In my experience they are just an arbitrary dictionary word with no relevance to anything, let alone the location/site/area, it's just an identifier.

bhousel commented:
> In my experience they are just an arbitrary dictionary word with no relevance to anything, let alone the location/site/area, it's just an identifier.

The names are funny but I actually think they are prominently displayed in the app and in real life.. If you go to the location, it says that you are at "Nunzio" or "Pail".  For example Google Maps shows this:

<img width="1267" alt="Screen Shot 2021-02-16 at 10 27 06 AM" src="https://user-images.githubusercontent.com/38784/108087561-92083d80-7045-11eb-8ac8-249219504fb9.png">

tas50 commented:
Another one that might be a bit too aggressive. This walmart was tagged as a supemarket and a pharmacy, but named for the supermarket:

![image](https://user-images.githubusercontent.com/1015200/108585170-b2f9b880-72fb-11eb-9225-b86e4e384247.png)

peternewman commented:
> > In my experience they are just an arbitrary dictionary word with no relevance to anything, let alone the location/site/area, it's just an identifier.
> 
> The names are funny but I actually think they are prominently displayed in the app and in real life.. If you go to the location, it says that you are at "Nunzio" or "Pail". For example Google Maps shows this:

Yes sorry, I didn't word that very well. I'd agree "Pail" is the most important way to know you're at the right Amazon Locker, but you need a map location to do that, you can't infer it's in Plainfield from the name (compared to say a supermarket where it would be "Costco Plainfield" for example).

Is there a beta site where we can start test-driving this @bhousel or do we need to do stuff locally to test it?

bhousel commented:
> Is there a beta site where we can start test-driving this @bhousel or do we need to do stuff locally to test it?

Yeah I would like this too - I think I can put something together soon.

I've been steadily picking my way though https://github.com/osmlab/name-suggestion-index/issues/4906 , which will give the iD validator a lot more flexibility in how it treats named things.

bhousel committed:
Move all of NSI into a service, rewrite matcher code

bhousel committed:
When displaying a preset image, use display:none for siblings
visibility:hidden still preserves space, which would uncenter the image in some cases

bhousel committed:
Unsquish the issue messages by adding more side padding

bhousel commented:
I spent a bunch of time on the NSI internals and we now have a lot more flexibility in how `name` and other tags are handled. 

I just tested some of the things that have come up in the testing and the latest version is looking pretty great 👍 :
- [x] Banner Goldfield Medical Center - don't overwrite `emergency` tag
- [x] Amazon Locker - allow custom `name`
- [x] Apple Store - allow custom `name`
- [x] Greene King - allow custom `name` 

This one is still tricky
- [x] Walmart Neighborhood Market + `amenity=pharmacy`

The validator is now able to prioritize the name and supermarket, but it does suggest to remove the `amenity=pharmacy` tag.  I sort of understand why, and I agree with the suggestion that the pharmacy tags should not be combined with the shop tags, but I might put something special into the code to avoid surprises like this, since this is probably a widespread mapping practice.


<img width="250" alt="Screen Shot 2021-03-03 at 2 48 11 PM" src="https://user-images.githubusercontent.com/38784/109863379-903a9e80-7c2f-11eb-8804-2ee9f1bc5b6a.png">

I'll try to get a version of this up for people to play around with asap.

tas50 commented:
Removing double tags might surprise some folks, but it's really the right thing to do for the map in general. I see a lot of combined drugstore/pharms and convenience store / gas stations. That's now how they should get tagged in OSM and it doesn't even align with how Apple or Google maps things. Having the editor provide another nudge to get users off the combined tagging model is a plus in my eyes.

bhousel commented:
I deployed this branch with the preview NSI v5 code to:  http://www.ideditor.org/nsi-v5/
Try it out!

UKChris-osm commented:
When trying a couple of samples in [South West of England](http://www.ideditor.org/nsi-v5/#background=Bing&disable_features=boundaries&map=16.80/50.97205/-3.18294):

- Typing "_Dev_" will suggest [Devon and Somerset Fire and Rescue Service](https://nsi.guide/index.html?t=operators&k=amenity&v=fire_station#devonandsomersetfireandrescueservice-ab1fb5), as expected, but when typing in "_fire station_" didn't suggest any NSI fire station entries at all. When typing "_Poli_" then Avon & Somerset police did show up, as did South West Ambulance when typing in "_Amb_".
- When typing in "_Emergency Phone_", Highways England didn't show up at all, which I was expecting.

Is this because the matcher only works on names, and not types?

bhousel committed:
For some names, consider splitting `name` into `name` and `branch`..

bhousel committed:
Don't offer upgrades to dissolved items

bhousel committed:
Improvements to name gathering
- Actually bail out properly if namelike values contain semicolon ';'
- Prioritize the canonical tags like `brand:wikidata`, `operator:wikidata`
- Don't insert a name into 'alternate' set if it's already in 'primary' set

bhousel committed:
Only match alternate `amenity/yes` if it actually is tagged that way

bhousel commented:
> Is this because the matcher only works on names, and not types?

Hey @UKChris-osm - yes what you're seeing is because of how iD preset matching works.  It is set to return the core presets before the suggestion presets.

>  but when typing in "fire station" didn't suggest any NSI fire station entries at all.

I actually do see these but they are far down in the list below around 15 core presets.  There are just a lot of presets that match "fire"

> When typing in "Emergency Phone", Highways England didn't show up at all, which I was expecting.

This is weird to me too - I'll play around and see if I can get the iD preset matcher to return more relevant results.

ALSO - I just pushed the latest version to http://ideditor.org/nsi-v5/ for testing, which improves a bunch of things.
(we're tracking some issues both here and on https://github.com/osmlab/name-suggestion-index/issues/4543)

bhousel committed:
Include nsi_dissolved in test setup

UKChris-osm commented:
> I actually do see these but they are far down in the list below around 15 core presets. There are just a lot of presets that match "fire".

Ah yes, I see now they are listed, just further down off screen. Is there scope to make the suggestions a two step process, so if I were to select the generic "_fire station_" iD preset, that the fire stations within the NSI would then be suggested in a secondary way?

Also, when I added a node as a generic fire station, I typed in "_Devon and Somerset_" as the operator, then clicked away, then clicked back, but no suggestion came up suggesting Devon and Somerset Fire and Rescue Service.

bhousel commented:
> Ah yes, I see now they are listed, just further down off screen. Is there scope to make the suggestions a two step process, so if I were to select the generic "_fire station_" iD preset, that the fire stations within the NSI would then be suggested in a secondary way?

I don't think we could introduce 2 steps of preset selection, but I can try to make the results more relevant.

> Also, when I added a node as a generic fire station, I typed in "_Devon and Somerset_" as the operator, then clicked away, then clicked back, but no suggestion came up suggesting Devon and Somerset Fire and Rescue Service.

The validator would suggest it if there is a `matchName` or some kind of alternate name like "Devon and Somerset".  
It needs to match something in NSI, by [key, value, name, location].

bhousel commented:
>> The validator is now able to prioritize the name and supermarket, but it does suggest to remove the amenity=pharmacy tag. I sort of understand why, and I agree with the suggestion that the pharmacy tags should not be combined with the shop tags, but I might put something special into the code to avoid surprises like this, since this is probably a widespread mapping practice.
>
> Removing double tags might surprise some folks, but it's really the right thing to do for the map in general. I see a lot of combined drugstore/pharms and convenience store / gas stations. That's now how they should get tagged in OSM and it doesn't even align with how Apple or Google maps things. Having the editor provide another nudge to get users off the combined tagging model is a plus in my eyes.

I agree 100% that it's the right thing to do, but I took out the code anyway because we have run into it a few times now:
- removing `emergency=yes` from a hospital [above](https://github.com/openstreetmap/iD/pull/8305#issuecomment-778707926) (because `emergency` is also a toplevel tag for phones, ambulance stations, etc), 
- removing `man_made=surveillance` tag [here](https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-791913156) (because `man_made` is also a toplevel tag for flagpoles)
- removing `amenity=pharmacy` from Walmart Neighborhood Market [above](https://github.com/openstreetmap/iD/pull/8305#issuecomment-782565876) (because `amenity` and `shop` are both toplevel tags)

So for now, I think we have to just leave these in.  I adjusted the code to only remove unspecific tags like `amenity=yes`, or `shop=yes` (but not `emergency=yes` 🙄 ).

I'll try to think about whether someday we can find a way to nudge people towards following the ["one feature, one element" rule](https://wiki.openstreetmap.org/wiki/One_feature,_one_OSM_element), in a way that doesn't surprise or remove data..

This is really the biggest challenge with OSM's free tagging model: tags are used to identify _what a feature is_, but occasionally those same tags are used to identify _something a feature has_.

bhousel committed:
Be less aggressive about removing toplevel tags
(re: https://github.com/openstreetmap/iD/pull/8305#issuecomment-792871448)

bhousel committed:
Support a more verbose format for listing issues
Includes the feature type in the label
(re: https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-791919134)

bhousel committed:
Remember user's preference for expanding issue-info section
This shows the tag diff, and other helpful information about the detected issues.
It is expanded by default now
(closes #6408, closes #8143)

bhousel commented:
I pushed the latest work to https://ideditor.org/nsi-v5 - let me know what you think!

peternewman commented:
> In future the iD upgrade could also populate the branch/ref field where it's currently only included in the name.

_Originally posted by @peternewman in https://github.com/osmlab/name-suggestion-index/issues/4906#issuecomment-782896882_

I see it's now offering this thanks @bhousel e.g. for https://www.openstreetmap.org/way/217820293

It's currently also proposing to strip it from the name. Is there some config for when it appears on the store itself too (I don't know if it does in this case)/how's that all being handled?

bhousel committed:
Adjust verbose utilDisplayLabel call, use for more validation messages
(re: https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-794527259)

bhousel committed:
Be smarter about identifying what tree an osm feature might be in
Needed this after I added `highway=bus_stop` to the transit tree

bhousel committed:
update imagery

bhousel committed:
Make sure a name is either primary or alternate (can't be both)

bhousel committed:
If we match a generic name, stop looking
This is for "nextbike"
(re: https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-796126041)

bhousel committed:
Create the categories like the presets
The lack of locationSet was causing them to not show up, which indirectly
broke the tutorial at the step where user needs to pick the road category.

bhousel committed:
Allow missing locationSetID on presets, fields, categories
This simplifies the code a bit, just treat it as worldwide if missing

bhousel committed:
Revise name/branch splitting code
This now breaks the name into fragments and reruns the fragments against the NSi matcher
rather than using flaky regular expressions. Has a few advantages:
(re: https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-797577999)

UKChris-osm commented:
If a user selects the suggested tags from the NSI, should some sort of reference be included in the source tag, either on the element itself, or within the changeset log?

Adamant36 commented:
"If a user selects the suggested tags from the NSI, should some sort of reference be included in the source tag, either on the element itself, or within the changeset log?" 

I've thought about that myself. At least with it being mentioned in a changeset comment (doing it on the element itself is a bad idea IMO). 

The problem is that the "anti-automated edit" crowed in OSM is way to bullish in their opposition to anything having to do with iD Editor or "automated edits", there isn't clear definition of what constitutes ones and the people who are ardently against them aren't willing to provide either, and so declaring how an edit was made if it was done in iD Editor with the NSI a recipe for targeting/harassment/and arbitrary blocks/reverts that are based more on "automated edit" fear mongering then if the edit was actually correct or not. 

Other software has presets and don't have to declare them as the source in changeset comments either. There isn't anything special about the NSI that it should have to compared to any other software that does similar things. 

The "source" could just as easily be "I verified this myself and I used the preset after doing so" anyway.

bhousel committed:
Switch to published NSI v5  :tada:

bhousel committed:
Improve logic for matching name fragments like TUI ReiseCenter
(re: https://github.com/osmlab/name-suggestion-index/issues/4543#issuecomment-799732746)

bhousel committed:
Location-aware preset matching

bhousel committed:
Add guard code in `locationsAt`, for testing entities with invalid loc

bhousel commented:
> It's currently also proposing to strip it from the name. Is there some config for when it appears on the store itself too (I don't know if it does in this case)/how's that all being handled?

Hey @peterneubauer - I hadn't replied to this before because I was still trying a few different approaches in code, but what I settled on was basically:
- the `name` tag is special and the code that looks for name candidates will also look for branches in the `name` tag
- it does it by breaking the name value into pieces and trying to match those fragments longest to shortest until it finds a name:
  - "TUI ReiseCenter Neuss Innenstadt" - no match
  - "TUI ReiseCenter Neuss" - no match
  - "TUI ReiseCenter" - match!  ("Neuss Innenstadt" is suggested as the `branch`)

This approach fails in a few interesting ways, but these are things I think we can handle in the name-suggestion-index with either separate entries for store formats or matchNames for alternate names.  I'm really interested in testing more of these.
- "Gap Kids", "Gap Factory Outlet" - (in these cases "Kids" and "Factory Outlet" are not branches).
- "KFC / Taco Bell" (It thinks "Taco Bell" is the branch, need to fix)

bhousel commented:
> Other software has presets and don't have to declare them as the source in changeset comments either. There isn't anything special about the NSI that it should have to compared to any other software that does similar things.

Yes this is a good point - NSI is really a preset pack + a separate matcher for alternate names.  I don't think of it as a "source" in the OSM sense.

bhousel committed:
More sophisticated name/branch splitting
(re: https://github.com/openstreetmap/iD/pull/8305#issuecomment-805454952)

bhousel commented:
I've been doing a bunch of testing with this the past few days, and I think it's about as good as it will get..
NSI v5 has been published, and this branch now pulls the latest compatible release of that.
I'd love for more people to test here and let me know if you notice any issues!  - https://ideditor.org/nsi-v5

lkernan commented:
https://ideditor.org/nsi-v5 seems to have stopped keeping up with changes to NSI somewhere around the 5.0.20210326 release commit.

UKChris-osm commented:
> https://ideditor.org/nsi-v5 seems to have stopped keeping up with changes to NSI somewhere around the 5.0.20210326 release commit.

I think @bhousel has changed this PR to use the "_released_" version of the NSI, rather than the pre-release that is generated when "_npm run dist_" is run.

bhousel commented:
> https://ideditor.org/nsi-v5 seems to have stopped keeping up with changes to NSI somewhere around the 5.0.20210326 release commit.

Yes!  20210326 is the latest published release. That's 2 days ago..

<img width="328" alt="Screen Shot 2021-03-28 at 7 51 18 AM" src="https://user-images.githubusercontent.com/38784/112751167-7d02c080-8f9a-11eb-85f7-40abd7b4f053.png">

And that is the version of the `nsi.min.json` that iD fetches from the CDN:
![Screen Shot 2021-03-28 at 7 51 24 AM](https://user-images.githubusercontent.com/38784/112751172-7ffdb100-8f9a-11eb-97e4-51c8db17cf45.png)

So that's actually pretty awesome that iD can do this now.  It won't pull the latest code from the `main` branch, as that is often broken, but we can publish updates very frequently now.

lkernan commented:
> > https://ideditor.org/nsi-v5 seems to have stopped keeping up with changes to NSI somewhere around the 5.0.20210326 release commit.
> 
> I think @bhousel has changed this PR to use the "_released_" version of the NSI, rather than the pre-release that is generated when "_npm run dist_" is run.

Oops sorry, that’s what confused me.

bhousel committed:
Upgraded to location-conflation v0.8.0

UKChris-osm commented:
@lkernan A new release is now available to test which includes your updates to the NSI 👍

kymckay reviewed (commented):
I've gone through the code closely and it all looks good to me, plus the testing introduced for the new locations module is very reassuring. Strong use of promises and idle callbacks 👍

Only thing I'm curious about is in the use of `setTimeout` in the tests, I see the PhantomJS comment and assume the problem is some kind of race condition where the test "server" can terminate before those async tests ever resolve?

bhousel commented:
> Only thing I'm curious about is in the use of `setTimeout` in the tests, I see the PhantomJS comment and assume the problem is some kind of race condition where the test "server" can terminate before those async tests ever resolve?

It looks like I added those delays into the test code elsewhere in this commit: https://github.com/openstreetmap/iD/commit/c481f90e7df21bdb464fbfa34249b2b7fae7c1f7

"This also includes a bunch of tweaks to make the tests work
in both PhantomJS and modern browsers like Chrome.

Basically - introduce some more async into the test code so that
the coreData.get promise is guaranteed to settle.  Because in
PhantomJS the promise is polyfilled, and Chrome it's native, they
work slightly differently."

kymckay reviewed (approved):
Ah of course, I hadn't considered the async polyfill behaviour differing from the real thing - core-js really is a magical thing!

In that case, the code gets logical/semantic approval from me. Haven't played too much with the preview, but it definitely works and seems there are no lingering significant design concerns from those who have spent more time with it.

Just throwing some weight in towards getting this beast merged before it becomes a truly long-life branch. 😄

bhousel committed:
bump versions

bhousel committed:
Upgrade to name-suggestion-index v6
Also bump country-coder and location-conflation

mbrzakovic commented on modules/core/context.js:
```diff
@@ -557,7 +557,7 @@ export function coreContext() {
       Object.values(services).forEach(service => {
         if (service && typeof service.init === 'function') {
-          service.init();
+          service.init(context);
```
Is this needed? I can't seem to find service that requires context.

bhousel replied:
probably not needed, I think all the services do not depend on context these days.

mbrzakovic commented on modules/core/locations.js:
```diff
@@ -0,0 +1,267 @@
+  function processQueue() {
+    if (!_queue.length) return Promise.resolve();
+
+    // console.log(`queue length ${_queue.length}`);
```
Few redundant debug comments.

bhousel replied:
They are great for debugging tho

mbrzakovic replied:
ok :)

mbrzakovic commented on modules/core/locations.js:
```diff
@@ -0,0 +1,267 @@
+        });
+        _deferred.add(handle);
+      })
+      .then(() => processQueue());
```
Nice trick 👌 (no-op)

mbrzakovic commented on modules/core/locations.js:
```diff
@@ -0,0 +1,267 @@
+      if (!locationSet) {
+        throw new Error('object missing locationSet property');
+      }
+      const resolved = _loco.resolveLocationSet(locationSet);
```
I did some debugging, it seems that Presets_presets objects with locationSet that only has exclude (e.g.amenity/pub/irish in [preset_presets](https://github.com/openstreetmap/id-tagging-schema/blob/main/dist/presets.json)) will be treated as errors due to missing 'includes' and handled here as Q2. Is this expected?

bhousel replied:
That is not expected, I think the json schema for these should require each locationSet to have an `include` property.  But I will check for a missing `include` and set it to `['Q2']` to mimic the old behavior of the presets (available worldwide).

mbrzakovic commented on modules/core/locations.js:
```diff
@@ -0,0 +1,267 @@
+      if (!resolved.feature.geometry.coordinates.length || !resolved.feature.properties.area) {
+        throw new Error(`locationSet ${locationSetID} resolves to an empty feature.`);
+      }
+      if (!_resolvedFeatures[locationSetID]) {  // First time seeing this locationSet feature
```
I am curious:  _loco has internal caching, why is another _resolvedFeatures cache required?

bhousel replied:
Good question, and I don't really remember why I wrote it this way.  
I think it might just have been defensive coding - this code makes a copy of the Object in `_resolvedFeatures` for whatever iD wants to do with it, instead of passing around references to the `_loco` internally cached Objects.

mbrzakovic replied:
Got it. no-op for now as impact is low and priority is pushing this change through

mbrzakovic commented on modules/services/nsi.js:
```diff
@@ -0,0 +1,668 @@
+  let t = identifyTree(tags);
+  if (!t) return empty;
+
+  if (t === 'transit') {
```
This seems very similar to: https://github.com/osmlab/name-suggestion-index/blob/main/dist/trees.json
Why not just use that data?

bhousel replied:
I don't really remember, but I think it was because:
-  in NSI it's easy to match what tree a feature belongs to, because the data is organized that way... 
-  in iD, it's a lot more complicated because it can only rely on a feature's tags, and those tags need to be matched in a certain order in some cases - primary before alternate + the exception for flag countries, and a feature can possibly match multiple trees.

mbrzakovic commented on modules/services/nsi.js:
```diff
@@ -0,0 +1,668 @@
+//
+function setNsiSources() {
+  const sources = {
+    'nsi_data': 'https://cdn.jsdelivr.net/npm/name-suggestion-index@6.0/dist/nsi.min.json',
```
Just to double-check. The url (version part) should be changed manually on each package version bump?
If so, are there any suggestions for automatization of this part? Perhaps we could at least document these kind of updates. (no-op)

bhousel replied:
I don't think we'd want to automate it.. The CDN will already serve the latest file that matches the requested [semver](https://semver.org/) number, so 6.0.xxxxxx.

I've been following semver on the NSI side, so if there is a major bump from 5 to 6, that means there is some incompatibility in the code that need to be addressed.  I always put stuff like this in the [changelog](https://github.com/osmlab/name-suggestion-index/blob/main/CHANGELOG.md).

Basically my thinking is:
- major bump - maybe some code needs to be rewritten, maybe data file format is different
- minor bump - code should still work, file format same maybe with new properties (NSI might get a new tree)
- patch bump - code is same, file format is same, some data is new. (NSI publishes a new patch about every week).

mbrzakovic replied:
Got it. I agree.
Ok then I have only this small observation: When new major/minor packages version becomes available, maintainer would have to bump version in both package.json and url links. It would be more secure to have one singular place with version and for that info to propagate automatically wherever its needed.

bhousel replied:
Oh yeah! I didn't understand before what you meant by automatization originally but I do now.  I think I can make a small improvement here so that this code constructs the URL to have the appropriate version number that matches whatever `package.json` has.

bhousel replied:
I did this, check out 8db1c1f7f

bhousel replied:
Can use this trick for a handful of other data files too, like osm-community-index, or the presets

mbrzakovic reviewed (changes requested):
It was amazing to see presets, nsi and oci common locationSet functionality coming together.
This is surely a big improvement and sets a strong foundation to make very easy additions of locationSet specifics to any of these packages. Also, its great that several issues have been addressed.

Perhaps we could include this concept in Arhitecture documentation as it has impact across multiple files and its not that easy to digest.

I went through the code and added few minor comments and some curious questions.
Lets go through them and merge this change.

mbrzakovic commented:
I saw data/imagery.json being updated in this PR.
Can you please provide info on how are these imagery sources gathered?

bhousel commented:
> I saw data/imagery.json being updated in this PR.
> Can you please provide info on how are these imagery sources gathered?

Yeah that's just ELI.  Anyone who works on iD needs to update it manually because ELI doesn't actually publish code with version numbers.  (so npm or package-lock can't work with it).  Everyone who installs iD is going to get whatever version of ELI happened to exist at the time - there will always be changes in `imagery.json`.  See #7425 or PM me for details.

bhousel committed:
If locationSet is missing `include`, default to worldwide include
https://github.com/openstreetmap/iD/pull/8305#discussion_r662344647

bhousel committed:
Construct URL to match version number that package.json has
https://github.com/openstreetmap/iD/pull/8305#discussion_r663037548

tas50 commented:
@bhousel is there still a URL floating around with this branch?

bhousel committed:
Remove unnecessary context argument

bhousel commented:
> @bhousel is there still a URL floating around with this branch?

Yes I've kept https://ideditor.org/nsi-v5 updated to pull latest released NSI

mbrzakovic committed:
Merge conflicts resolved

lgommans commented:
Why did this close #6408? It changed nothing in terms of seeing what you're changing, iD still invites people to blindly apply changes under their name.
